File: Python3/프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    infos = {}
    for i in range(len(genres)):
        genre = genres[i]
        if infos.get(genre): # genre가 존재하는 경우
            infos[genre][0] += plays[i]
            infos[genre][1].append((plays[i], i)) # play와 index 삽입
        else: # genre가 존재하지 않는 경우
            infos[genre] = [plays[i], [(plays[i], i)]]
    
    infos = dict(sorted(infos.items(), key = lambda x:x[1][0], reverse=True))
    for k, v in infos.items():
        play = v[1]
        print(play)
        play.sort(key = lambda x:x[1])
        play.sort(key = lambda x:x[0], reverse=True)
        print(play)
        length = 2 if len(play)>=2 else len(play)
        for i in range(length):
            answer.append(play[i][1])
            
    return answer

File: Java/프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    
    static class Music{
        String genre;
        int play;
        int idx;
        
        public Music(String genre, int play, int idx){
            this.genre = genre;
            this.play = play;
            this.idx = idx;
        }
    }
    
    public int[] solution(String[] genres, int[] plays) {
        
        ArrayList<Integer> answer = new ArrayList<>();
        HashMap<String, Integer> map = new HashMap<>();
        for(int i=0;i<genres.length;i++){
            map.put(genres[i], map.getOrDefault(genres[i], 0)+plays[i]);
        }
        
        ArrayList<String> arr = new ArrayList<>();
        for(String g:map.keySet()){
            arr.add(g);
        }
        arr.sort((g1, g2)->map.get(g2)-map.get(g1));
        
        ArrayList<Music> music = new ArrayList<>();
        for(String g:arr){
            ArrayList<Music> genre_music = new ArrayList<>(); // 장르가 동일한 music list
            for(int i=0; i<genres.length; i++){
                if (genres[i].equals(g)){ // 현재 장르와 동일한 경우
                    genre_music.add(new Music(g, plays[i], i));
                }
            }
            genre_music.sort((m1, m2)->m2.play-m1.play); // play 내림차순으로 정렬
            answer.add(genre_music.get(0).idx);
            if(genre_music.size()>1){
                answer.add(genre_music.get(1).idx);
            }
        }
        
        // System.out.println(answer);
        return answer.stream().mapToInt(i->i).toArray();
    }
}